internal/worker/concurrency_auto_scaler_test.go

```diff
@@ -0,0 +1,392 @@
+	return NewConcurrencyAutoScaler(ConcurrencyAutoScalerInput{
+		Concurrency: &ConcurrencyLimit{
+			PollerPermit: NewResizablePermit(100),
+			TaskPermit:   NewResizablePermit(1000),
+		},
+		Cooldown:                 2 * testTickTime,
+		Tick:                     testTickTime,
+		PollerMaxCount:           200,
+		PollerMinCount:           50,
+		Logger:                   logger,
+		Scope:                    tally.NoopScope,
+		Clock:                    clock,
+		PollerWaitTimeLowerBound: 16 * time.Millisecond,
+		PollerWaitTimeUpperBound: 256 * time.Millisecond,
+	})
+}
+
+func TestConcurrencyAutoScaler(t *testing.T) {
+
+	type eventLog struct {
+		eventType   autoScalerEvent
+		enabled     bool
+		pollerQuota int64
+		time        string
+	}
+
+	for _, tt := range []struct {
+		name               string
+		pollAutoConfigHint []*shared.AutoConfigHint
+		expectedEvents     []eventLog
+	}{
+		{
+			"start and stop immediately",
+			[]*shared.AutoConfigHint{},
+			[]eventLog{
+				{autoScalerEventStart, false, 100, "00:00:00"},
+				{autoScalerEventStop, false, 100, "00:00:00"},
+			},
+		},
+		{
+			"just enough pollers",
+			[]*shared.AutoConfigHint{
+				{common.PtrOf(true), common.PtrOf(int64(16))}, // <- tick, in cool down
+				{common.PtrOf(true), common.PtrOf(int64(16))}, // <- tick, no update
+			},
+			[]eventLog{
+				{autoScalerEventStart, false, 100, "00:00:00"},
+				{autoScalerEventEnable, true, 100, "00:00:00"},
+				{autoScalerEventPollerSkipUpdateCooldown, true, 100, "00:00:01"},
+				{autoScalerEventPollerSkipUpdateNoChange, true, 100, "00:00:02"},
+				{autoScalerEventStop, true, 100, "00:00:02"},
+			},
+		},
+		{
+			"poller slightly idle but no change",
+			[]*shared.AutoConfigHint{
+				{common.PtrOf(true), common.PtrOf(int64(100))}, // <- tick, in cool down
+				{common.PtrOf(true), common.PtrOf(int64(100))}, // <- tick, no update
+			},
+			[]eventLog{
+				{autoScalerEventStart, false, 100, "00:00:00"},
+				{autoScalerEventEnable, true, 100, "00:00:00"},
+				{autoScalerEventPollerSkipUpdateCooldown, true, 100, "00:00:01"},
+				{autoScalerEventPollerSkipUpdateNoChange, true, 100, "00:00:02"},
+				{autoScalerEventStop, true, 100, "00:00:02"},
+			},
+		},
+		{
+			"busy pollers",
+			[]*shared.AutoConfigHint{
+				{common.PtrOf(true), common.PtrOf(int64(10))}, // <- tick, in cool down
+				{common.PtrOf(true), common.PtrOf(int64(10))}, // <- tick, scale up
+			},
+			[]eventLog{
+				{autoScalerEventStart, false, 100, "00:00:00"},
+				{autoScalerEventEnable, true, 100, "00:00:00"},
+				{autoScalerEventPollerSkipUpdateCooldown, true, 100, "00:00:01"},
+				{autoScalerEventPollerScaleUp, true, 116, "00:00:02"},
+				{autoScalerEventStop, true, 116, "00:00:02"},
+			},
+		},
+		{
+			"busy pollers, scale up to maximum",
+			[]*shared.AutoConfigHint{
+				{common.PtrOf(true), common.PtrOf(int64(0))}, // <- tick, in cool down
+				{common.PtrOf(true), common.PtrOf(int64(0))}, // <- tick, scale down to minimum
```

Comment:
```suggestion
				{common.PtrOf(true), common.PtrOf(int64(0))}, // <- tick, scale up significantly
```

Reply:
I'll keep the maximum case but will change the test cases below